bookdown/01-mlr3-basics.Rmd

```diff
@@ -24,29 +24,27 @@ The `r mlr_pkg("mlr3")` package provides R6 classes for the these essential buil
-* Classes usually expose some methods to
-  * retrieve information,
-  * perform some action,
-  * change the state of the object.
-
-  For example, `foo$inc()` might increment the value of `bar` in `foo` by `+1`, and hence changes the state of the object.
-* Classes have public and private parts of their interface, regarding methods and member variables.
-  We as outside users can only access the public parts and do not have to worry about the private ones.
-* R6 classes are implemented via environments / references.
-  This means that R6 variables do not actually hold the object itself but rather a reference (pointer) to them.
-  This means that `foo2 = foo` does not copy the object, but only copies the reference, which now points to the object.
-* If you really want to fully copy an object call `foo$clone()`, possibly `foo$clone(deep = TRUE)`.
-
-If you want more details, have a look at the [R6 vignettes](https://r6.r-lib.org/).
+## Quick R6 Intro for Beginners
+
+R6 is one of R's more recent dialects for object-oriented programming (OO).
+It addresses shortcomings of earlier OO implementations in R, such as S3, which we used in `r mlr_pkg("mlr")`.
+If you are not familiar with OO programming, it might take a bit to find your way around, but if you have done OO in another language, R6 should feel familiar.
+We focus on the parts of R6 that you need to know to use `r mlr_pkg("mlr3")` here without going into further detail.
+
+* Objects are created by calling the constructor of an `R6Class()` object.
+  For example `foo = Foo$new(bar = 1)` creates a new object of class `Foo`, setting the `bar` argument of the constructor to `1`.
+* Classes have mutable state which is encapsulated in the member variables.
+  Continuing with the previous example, assuming that the constructor sets the public member variable `bar` to `1`, we can access its value through the dollar operator: `foo$bar` or `foo$bar = 2`.
+* In addition to member variables, objects expose methods that may allow to inspect the object's state or retrieve information and perform an action that may change the internal state of the object.
+  As an example, the `$train` method of a learner changes the internal state by producing and storing a trained model.
+* Objects can have public and private member variables and methods.
+  As a user of `r mlr_pkg("mlr3")`, you can only access the public variables and methods -- the private are only relevant if you want to change or extend `r mlr_pkg("mlr3")`.
+* R6 variables are references to the actual object in the environment; they do not hold the object itself.
+  For example `foo2 = foo` does not create a copy of `foo` and store it in `foo2`, but only the reference to the actual object.
+  `foo` and `foo2` refer to the same object, and setting `foo$bar = 3` will also change `foo2$bar` to `3`.
+* To copy an object, use the `$clone()` method; use the `deep = TRUE` argument for nested objects, for example `foo2 = foo$clone(deep = TRUE)`.
```

Comment:
This should go before the cloning examples above.

Reply:
The examples above are not cloning but copying of a reference.